### PR TITLE
Parallel tpetra scatter gather

### PR DESCRIPTION
**Pull request by bathmatt**
I sped up several panzer evaluators gather/scatter and vector to scalar, by using parallel fors.  

@trilinos/panzer 

Suggesting either @jmgate or @rppawlo 

## Motivation and Context

This gained a factor of 5 in runtime for the EMPIRE application

## How Has This Been Tested?

Using gcc/opt/debug with openMP I got 
100% tests passed, 0 tests failed out of 114

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [ ] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/panzer/disc-fe/src/evaluators/Panzer_VectorToScalar_impl.hpp [hunk 1]
```diff
@@ -92,19 +92,18 @@ PHX_EVALUATE_FIELDS(VectorToScalar,workset)
...
-      
-	scalar_fields[sc](cell,pt) = vector_field(cell,pt,sc);
-
+  // Need local copies for cuda, *this is not usable
+  auto local_vector_field = vector_field;
+  // Loop over scalars
+  for (std::size_t sc = 0; sc < scalar_fields.size(); ++sc) {
+    auto local_scalar_field = scalar_fields[sc];
+    // Loop over cells
+    Kokkos::parallel_for(workset.num_cells, KOKKOS_LAMBDA (const index_t cell) {
```
Comment (rppawlo) [inline packages/panzer/disc-fe/src/evaluators/Panzer_VectorToScalar_impl.hpp]: I think this is the first kokkos lambda use in panzer. All other kokkos objects were functors. We might want to add a cmake check to make sure that Kokkos_ENABLE_Cuda_Lambda is ON when building panzer for CUDA.

Comment (bathmatt) [reply packages/panzer/disc-fe/src/evaluators/Panzer_VectorToScalar_impl.hpp]: I can redo it w/o a lambda, are we going that way?

Comment (bathmatt) [reply packages/panzer/disc-fe/src/evaluators/Panzer_VectorToScalar_impl.hpp]: Plus, I was about to push a second fix to use a lambda in gather as well..

**Comment by bathmatt** [comment]
I updated this MR, does it automagically rerun the tests?

File: packages/panzer/disc-fe/src/evaluators/Panzer_VectorToScalar_impl.hpp (hunk 1 above)
Comment (rppawlo) [reply packages/panzer/disc-fe/src/evaluators/Panzer_VectorToScalar_impl.hpp]: Personally I want to use lambdas, so I would keep as is. I can add a cmake check later for this.

**Comment by rppawlo** [comment]
> I updated this MR, does it automagically rerun the tests?

yes

**Comment by jmgate** [comment]
Do you want to add the cmake check in this PR, or later?

**Comment by bathmatt** [comment]
@jmgate up to you on this...  I require it in EMPIRE and I thought tpetra requiree it, but I don't see any use of it in their cmake files.

**Comment by mhoemmen** [comment]
@bathmatt Kokkos' recent changes messed me all up; I need to relearn how they spell all those CMake options....

**Comment by bathmatt** [comment]
@mhoemmen does tpetra require cuda lambda?

**Comment by bathmatt** [comment]
Or, at least will it once you change the names to protect the innocent (kernels)?

**Comment by rppawlo** [comment]
> Do you want to add the cmake check in this PR, or later?

Let's  not hold @bathmatt up. We can add later if we find out that tpetra doesn't already require it.

**Comment by jmgate** [comment]
Well shoot&mdash;should've squashed.

**Comment by bathmatt** [comment]
I tested this on shiller with cuda and it passes all panzer tests..

**Comment by mhoemmen** [comment]
@bathmatt wrote:
> does tpetra require cuda lambda?

Currently yes.  Does this block your work?

Kokkos examples and tutorial are full of lambdas, so it feels like lambdas are the right thing to use.  However, if lambdas are bad, we can start rewriting code, if that needs to be done.

**Comment by bathmatt** [comment]
No, @mhoemmen that is good, I've started doing stuff like

auto local_copy = this->var;
lambda goes here {}

which gets around the this* issues.

**Pull request merged**